storage/shard.go

```diff
@@ -182,6 +187,22 @@ func (entry *dbShardEntry) decrementReaderWriterCount() {
 	atomic.AddInt32(&entry.curReadWriters, -1)
 }
 
+func (entry *dbShardEntry) needsIndexUpdate(writeTime time.Time) bool {
+	return atomic.LoadInt64(&entry.reverseIndex.nextWriteTimeNanos) < writeTime.UnixNano()
+}
+
+// ensure dbShardEntry satisfies the `onIndexSeries` interface.
+var _ onIndexSeries = &dbShardEntry{}
+
+func (entry *dbShardEntry) OnIndexSuccess(nextWriteTime time.Time) {
+	atomic.StoreInt64(&entry.reverseIndex.nextWriteTimeNanos, nextWriteTime.UnixNano())
+}
+
+func (entry *dbShardEntry) OnIndexFinalize() {
+	// indicate the index has released held reference for provided write
+	entry.decrementReaderWriterCount()
```

Comment:
Can we perhaps have a synchronous API to `OnIndexFinalize(...)` and we can keep it private too so others don't need to see it, i.e. `onIndexPrepare(...) { entry.incrementReaderWriterCount() }`?   That way its clear why you would decrement here.

Reply:
sure thing

Reply:
done